### ISSUE TITLE
Can we lose writes with LevelDB?

### ISSUE THREAD

**Issue by yaronyg** [Level/leveldown-mobile#10]
We are using leveldown-mobile and are trying to figure out if we can lose writes.

The scenario is that we get notified by Android or iOS that our app is about to be shut down. We get a short window of time to try to do clean up before we are ended.

Our understanding is that LevelDB uses an in memory cache to store information and that writes are done using the write() system file APIs. In which case even if our process is killed then no data should be lost because it’s the OS that has the write buffer, not the app. So in theory we shouldn’t need to worry about trying to do any kind of flush (not that LevelDB seems to support the concept anyway) nor should we worry about passing in the sync flag (and thus doing a fsync() not to mention obliterating our performance). 

Basically the only way we should be able to lose data (unless iOS or Android are doing something very bad) at the LevelDB layer is if the whole phone crashes. Does that sound right?

**Comment by obastemur** [comment]
Although I did the porting for mobile, I don't know the internals of LevelDB. What I understood from the question, members of (/cc @Level/leveldown) must have more to say on this. IMO iOS vs OSX Desktop doesn't make much difference. They both have persistent sync storage.

**Comment by yaronyg** [comment]
I'll close this here. I did send this request to the PouchDB folks (who are largely responsible for leveldown) and I haven't gotten a response.

**Comment by dominictarr** [comment]
@yaronyg that isn't right, leveldown was started by the level* team, and then the pouchdb people adopted it. they have certainly made a significant contribution to it, but leveldown started separately before pouchdb adopted it.

**Comment by yaronyg** [comment]
@dominictarr Thanks Dominic, I apologize for my misinformation. But where would be a good place to get an answer to my basic question regarding losing writes in LevelDB? Should I file a bug on the main LevelDB depot? Is there a mailing list I should mail? I really want to understand what kind of failure scenarios we are facing when we use LevelDB on mobile.

**Comment by nolanlawson** [comment]
Hey @yaronyg, just a quick comment - I did see your posts on [the mailing list](https://groups.google.com/forum/#!forum/pouchdb), but haven't had a chance to respond yet. Also @dominictarr is quite right that the PouchDB team doesn't really contain any LevelDB experts - all of us are just consumers of the API and can't really be considered core team members. Also it's worth noting that (if I'm not mistaken), the Node "level" community is not the same as the actual team at Google that builds/maintains LevelDB - you might want to ping [Joshua Bell](https://twitter.com/inexorabletash) about hardcore LevelDB questions.

Also kinda off-topic, but I just recently got PouchDB working over SQLite in Node, so there's a second option that's now available for the Thali project: https://github.com/pouchdb/pouchdb/issues/4915

**Comment by yaronyg** [comment]
The SQLite option is interesting although it's worth pointing out that Android has its own custom variant of SQLite that I believe is arbitrarily different than the main line. I know when I was working with CouchBase this proved to be a serious pain. Also, are we shoving attachments into SQLite?
